scripts/error-docgen.mjs

```diff
@@ -0,0 +1,256 @@
+# Do not make edits to it directly, they will be overwritten.
+# Instead, change this file: https://github.com/withastro/astro/blob/main/packages/astro/src/core/errors/errors-data.ts
+# Translators, please remove this note and the <DontEditWarning/> component.
+
+layout: ~/layouts/MainLayout.astro
+title: Error reference
+i18nReady: true
+githubURL: https://github.com/withastro/astro/blob/main/packages/astro/src/core/errors/errors-data.ts
+setup: |
+  import Since from '../../../components/Since.astro';
+  import DontEditWarning from '../../../components/DontEditWarning.astro';
+---
+
+<DontEditWarning />
+
+The following reference is a complete list of the errors you may encounter while using Astro. For additional assistance, including common pitfalls, please also see our [Troubleshooting Guide](/en/guides/troubleshooting/).
+
+`;
+
+const FOOTER = ``;
+
+export async function run() {
+	const compilerErrorData = { errors: [], jsdoc: [] }; // await getCompilerErrors();
+	const astroErrorData = await getAstroErrorsData();
+
+	let compilerResult = '';
+	for (const comment of compilerErrorData.jsdoc) {
+		if (comment.kind === 'heading') {
+			compilerResult += `## ${comment.name}\n\n`;
+			if (comment.description) {
+				compilerResult += comment.description.trim() + '\n\n';
+			}
+			continue;
+		}
+
+		const errorName = comment.meta.code.value;
+		compilerResult += [
+			`### ${padCode(compilerErrorData.errors[errorName])} - ${cleanCompilerError(errorName)}`,
+			'\n',
+		]
+			.filter((l) => l !== undefined)
+			.join('\n');
+	}
+
+	let astroResult = '';
+	for (const comment of astroErrorData.jsdoc) {
+		if (comment.kind === 'heading') {
+			astroResult += `## ${comment.name}\n\n`;
+			if (comment.description) {
+				astroResult += comment.description.trim() + '\n\n';
+			}
+			continue;
+		}
+
+		if (comment.see && comment.see.length === 1) {
+			comment.see = comment.see[0].split('\n');
+		}
+
+		const cleanMessage = comment.tags.find((tag) => tag.title === 'message')?.value;
+		astroResult += [
+			`### ${sanitizeString(
+				astroErrorData.errors[comment.meta.code.name].title ?? comment.longname
+			)}`,
+			comment.deprecated
+				? [
+						``,
+						':::caution[Deprecated]',
+						typeof comment.deprecated === 'string'
+							? comment.deprecated
+							: 'This error cannot be emitted by Astro anymore',
+						':::',
+				  ]
+						.filter((l) => l !== undefined)
+						.join('\n')
+				: undefined,
+			``,
+			`> ${getMessage(
+				comment.longname,
+				astroErrorData.errors[comment.longname].message,
+				cleanMessage
+			)} (E${padCode(astroErrorData.errors[comment.longname].code)})\n`,
+			comment.description && `#### What went wrong?`,
+			comment.description && comment.description.trim(),
+			``,
+			comment.see
+				? `**See Also:**\n${comment.see.map((s) => `- ${s.replace('-', '')}`.trim()).join('\n')}`
+				: undefined,
+			`\n\n`,
+		]
+			.filter((l) => l !== undefined)
+			.join('\n');
```

Comment:
Doesn’t need to block this PR, but this is… fun to read 😅 

If there’s time, I’d love to see a refactored version that breaks this logic up a bit, but understand if that’s not possible.

Reply:
This is definitely possible, a lot of the structure is based on the other docgen scripts (which are also a tiny bit hard to read). Will do my best to make things clearer!